src/entt/entity/polymorphic.hpp

```diff
@@ -0,0 +1,554 @@
+        inline static Base convert(Derived ptr) ENTT_NOEXCEPT {
+            return static_cast<Base>(ptr);
+        }
+    };
+
+    /** @brief value to pointer conversion, workaround for if constexpr bug on some compilers */
+    template<typename T, typename = void>
+    struct value_to_ptr {
+        inline static T* convert(T& ref) ENTT_NOEXCEPT {
+            return std::addressof(ref);
+        }
+    };
+
+    /** @copydoc value_to_ptr */
+    template<typename T>
+    struct value_to_ptr<T, std::enable_if_t<std::is_pointer_v<T>>> {
+        inline static T convert(T ptr) ENTT_NOEXCEPT {
+            return ptr;
+        }
+    };
+
+public:
+    /** @brief entity type of the component pools */
+    using entity_type = Entity;
+    /** @brief component type */
+    using value_type = Type;
+    /** @brief type of the underlying pool container */
+    using pools_container = std::vector<poly_pool_holder_base<Entity>>;
+    /** @brief pool iterator type */
+    using pools_iterator = internal::converting_iterator<cast_pool_holder, typename pools_container::iterator>;
+    /** @brief const pool iterator type */
+    using const_pools_iterator = internal::converting_iterator<const cast_pool_holder, typename pools_container::const_iterator>;
+    /** @brief single entity component iterator type */
+    using component_iterator = internal::poly_components_iterator<pools_iterator>;
+    /** @brief const single entity component iterator type */
+    using const_component_iterator = internal::poly_components_iterator<const_pools_iterator>;
+
+    /**
+     * @brief From a given set, makes a poly_storage_holder to hold child type and convert to parent type.
+     * @tparam ChildType source type to convert from
+     * @param pool_ptr child (source) type pool
+     * @return poly_storage_holder to hold ChildType set and access it as Type
+     */
+    template<typename StorageType>
+    inline static poly_pool_holder<Entity, Type> make_pool_holder(StorageType* pool_ptr) ENTT_NOEXCEPT {
+        using BaseType = Type;
+        using DerivedType = typename StorageType::value_type;
+        static_assert(is_poly_type_v<DerivedType>);
+        static_assert(type_list_contains_v<poly_parent_types_t<DerivedType>, BaseType> || std::is_same_v<DerivedType, BaseType>);
+        static_assert(std::is_pointer_v<DerivedType> == std::is_pointer_v<BaseType>);
+
+        void* (*get)(void*, Entity entity) ENTT_NOEXCEPT = +[](void* pool, const Entity entity) ENTT_NOEXCEPT -> void* {
+            if (static_cast<StorageType*>(pool)->contains(entity)) {
+                // if entity is contained within the set
+                if constexpr(std::is_base_of_v<std::remove_pointer_t<BaseType>, std::remove_pointer_t<DerivedType>>) {
+                    // if base type is inherited from derived type, do pointer conversion
+                    return derived_to_base_ptr<BaseType, DerivedType>::convert(static_cast<StorageType*>(pool)->get(entity));
```

Comment:
This is the workaround for `if constexpr` bug on MSVC v141, mentioned in the comment above